docs/agent-development.md

```diff
@@ -384,3 +387,55 @@ Here's a list of the config options across all agents, their types, default valu
+
+To minimise the amount of work required by users, agents should aim to enable this feature by default. This excludes RUM, where there is a performance penalty.
+
+#### Caching
+
+As mentioned above, the server will cache config for each unique `service.name`, `service.environment` pair. The server will respond to config requests with two related response headers: [Etag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) and [Cache-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control).
+
+Agents should keep a record of the `Etag` value returned by the most successful config request, and then present it to future requests via the [If-None-Match](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match) header. If the config has not changed, the server will respond with 304 (Not Modified).
+
+The `Cache-Control` header should contain a `max-age` directive, specifying the amount of time (in seconds) the response should be considered "fresh". Agents should use this to decide how long to wait before requesting config again. The server will respond with a `Cache-Control` header even if the request fails.
+
+#### Dealing with errors
+
+Agents must deal with various error scenarios, including:
+
+ - 7.3 servers where the Kibana connection is not enabled (server responds with 403)
+ - 7.3 servers where the Kibana connection is enabled, but unavailable (server responds with 503)
+ - any other error (server responds with 5xx)
+ - pre-7.3 servers that don't support the config endpoint (server responds with 404)
+
+If the server responds with 403, agents should log the response at info level.
+
+If the server responds with any 5xx, agents should log at error level. In any case, the server _should_ respond with a Cache-Control header, as described in the section above, and agents should retry after the specified interval. If for whatever reason the server does not respond with that header, or it is invalid, agents should retry after 5 minutes.
+
+If the server responds with 404 (i.e. it's a pre-7.3 server that lacks the config endpoint), agents may choose to treat it the same as 403, or may choose to log only once and mark that server as not having support for config, and then never retry. Bear in mind that this would mean the application must be restarted if the server is upgraded to 7.3.
+
+If the agent does not recognise a config attribute, then it should log a warning such as:
+
+```
+Remote config failure. Unsupported config names: unknown_option, other_unknown_option
+```
+
+Note that in the initial implementation of this feature, not all config attributes will be supported by the APM UI or APM Server. Agents may choose to support only the attributes supported by the UI/server, or they may choose to accept additional attributes. This will enable them to work without change once additional config attributes are supported by the UI/server.
```

Comment:
It's unclear from this if the agent should log the above-mentioned warning in the case that the agent decided only to support the same config options that the APM UI currently supports, but it receives another config option that's technically known, but unsupported.

Currently, this is how the Node.js agent has implemented it. I.e. it accepts only `transaction_sample_rate`, and logs a warning for any other config option.

Reply:
I changed the wording slightly, and updated the example to use known config attributes that agents may not support dynamically updating in the initial implementation.